### PR TITLE
Move commit checker from Azure Pipelines to Github Actions

### PR DESCRIPTION
**Pull request by glx22**
Basic workflow but logs should be less hidden and easier to read.
It can probably be improved with annotations and other stuff.

### CODE REVIEW

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml]: please remove this newline

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml]: please switch to `@v2`, and add newlines between `- name` entries

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml]: `${GITHUB_BASE_REF}` (always {} your variables!)

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml:39]: Possibly we should move this to https://github.com/OpenTTD/actions, but that is future work

**Review by TrueBrain — COMMENTED** [review]
Minor things, more style; lgtm otherwise

**Comment by glx22** [comment]
Running version [here](https://github.com/glx22/OpenTTD/pull/2)

File: .github/workflows/commit-checker.yml
Comment (glx22) [inline .github/workflows/commit-checker.yml]: All of this is probably not needed as a merge is always on top of the target ;)

**Review by glx22 — COMMENTED** [review]
I found a way to get all the PR commits. First step is to get the latest commit of target branch. Then we can get commits only in the merge. Finally the commit check is done for the relevant commits.

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml]: please enable 'set -x', as it is impossible to see what it really did like this :D

I have to experiment a bit to see if this really does what you hope it does in several scenarios .. hard to estimate when looking at it like this

Comment (TrueBrain) [reply .github/workflows/commit-checker.yml]: Okay, this is a bit difficult to test what it does exactly etc. It seems to only fetch the (full) history of 'master', and not the release branches etc. So there is some speed-up. Of course ideal we want to only fetch the commits between master and pull-request, but that seems to be a difficult thing to do in git-world.

I do have to say that I find this to have a lot of complexity, of which I am not sure it always works. On the other hand I am tempted to make a pull-request in `actions/checkout`, adding the `negotiation-tip`. Puzzling :)

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml]: I do not understand the `HEAD^2` .. to me it sounds like it should be just `HEAD^` (as that is the merge commit). Why skip the next commit too?

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [reply .github/workflows/commit-checker.yml]: Okay, this is a bit difficult to test what it does exactly etc. It seems to only do what you hope it does when the pull-request is pretty recently rebased on master. Otherwise the full history is fetched, and it takes as long as just using the checkout action without comments.

Is that worth the complexity, I wonder?

Okay, more experimenting later, it is exactly that: it stops sending the logs when the `negotiation-tip` commit is reached or when it reaches the end of the tree. So this indeed only works for any pull-request that is rebased against `master` at the time of running this action. After that, it just fetches the full tree.

File: .github/workflows/commit-checker.yml
Comment (glx22) [reply .github/workflows/commit-checker.yml]: It's a merge commit, so there are more than one parent. `HEAD^` is the first parent, master branch. `HEAD^2` is the second parent, the PR branch.

Comment (TrueBrain) [reply .github/workflows/commit-checker.yml]: Okay, never knew. Also deserves a comment for sure :D But that means this fails, as the chances are pretty high for a pull-request to not be on `origin/master` (the tip), so this compare will not return any results. I guess that is why the rebase was there :) That somewhat makes sense now ..

You can also use `git merge-base` I guess to find the common point. But for this you need a higher fetch-depth on the `origin/master`, otherwise it won't find them.

All in all, I have to conclude that this really only works for a small subset of our PRs. Most people don't bother updating `master` that often, so pull requests are often behind one or more commits.

I tried to find other ways to do this without fetching the whole repository, but it seems non-trivial. Do you agree with my findings? :)

Comment (glx22) [reply .github/workflows/commit-checker.yml]: I totally agree :)
I'm also trying to find an elegant way to fetch more only if needed.

Comment (TrueBrain) [reply .github/workflows/commit-checker.yml]: There are gists that do this; but are they worth it? Complexity vs speed .. :D

I saw some gists where people just did a for-loop increasing `fetch-depth` until `merge-base` returned a result. Silly, but efficient I guess. Just many hits on the server .. not sure GitHub appreciates that :D

Comment (glx22) [reply .github/workflows/commit-checker.yml]: Yeah, I think I'll start with a reasonable depth (enough for most PRs) and then double the depth in a loop only if it's really needed (for rare huge PRs)

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml]: Lets put the initial depth on 4. That is the merge commit, the master, and 2 commits in the PR. Average PR has one commit, so that means we capture just more than average.

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml]: "actions/checkout did a merge checkout of the pull-request. As such, the first commit is the merge commit. This means that on HEAD^ is the base branch, and on HEAD^2 are the commits from the pull-request. We now check if those trees have a common parent. If not, we fetch a few more commits till we do. In result, the log between HEAD^ and HEAD^2 will be the commits in the pull-request", or something like that?

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml]: `while [ -z "$(git merge-base HEAD^ HEAD^2)" ]; do`
in general, please don't use `!` for non-numbers, and don't use the ` shell escape :D

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml]:
```
DEPTH=4
while ...
--deepen=${DEPTH}
DEPTH=$(( ${DEPTH} * 4 ))
```

So that means first 4 more commits, 16, 64, ...

Worst case it will be fetching 24000 commits.  So that would be around 6 calls. Not pretty, but .. I guess there is a balance to be found.

File: .github/workflows/commit-checker.yml
Comment (TrueBrain) [inline .github/workflows/commit-checker.yml]: As we explained this earlier, I am fine not repeating ourself :)

**Review by TrueBrain — APPROVED** [review]
LGTM! Let's fix things when we find cases where it breaks :D

**Review by LordAro — APPROVED** [review]

**Pull request merged**
